### DISCUSSION TITLE
Warp-like features

### DISCUSSION THREAD

**Discussion by Marin-Kitagawa** [microsoft/terminal#13146] [Ideas]
# Description of the new feature/enhancement

The terminal [warp](https://www.warp.dev) supports many features such as
1. Terminal Notebooks
2. Blocks
3. Workflows
4. Terminal used like a text editor
5. New features upcoming is documentation being accessible as Terminal notebooks

This will be very great if it is implemented for Windows Terminal. They only released it for Mac OS not even for Linux. Each of the commands are like blocks and hence it is easy to scroll and look up and also easy to copy the output. We can create workflows to store the commands and invoke them via search. 

These will be great features to incorporate.

**Comment by SnirBroshi** [comment]
This requires co-operation from the running shell, so I see 3 ways to do this:

1. From [How Warp Works](https://www.warp.dev/blog/how-warp-works):

    > Given all the terminal sees is characters in and characters out, trying to determine when a command is run is next to impossible using just the characters from the pseudoterminal.
    > Thankfully, most shells provide hooks for before the prompt is rendered (zsh calls this [precmd](https://zsh.sourceforge.io/Doc/Release/Functions.html)) and before a command is executed ([preexec](https://zsh.sourceforge.io/Doc/Release/Functions.html)). Zsh and Fish have built in support for these hooks. Though bash does not have built in support for these hooks, scripts like [bash-preexec](https://github.com/rcaloras/bash-preexec) exist to mimic the behavior of other shells. Using these hooks, we send a custom DCS ([Device Control String](https://vt100.net/docs/vt510-rm/chapter4.html#:~:text=Device%20control%20strings%20(DCS)%2C,for%20a%20device%20control%20string.)) from the running session to Warp. This DCS contains an encoded JSON string that includes metadata about the session that we want to render. Within Warp we can parse the DCS, deserialize the JSON, and create a new block within our data model.

    So when a session starts they run their own command which is hidden from the user to set up hooks that report back to them on each command.
    I think this is similar to how VSCode provides information about the running shell in its terminal (such as cwd), but I'm not sure.

2. Build a custom shell, like [Upterm](https://github.com/railsware/upterm) did, and have it work only with it.
This will probably lead to the best experience but it's a shame users won't be able to use their shell with the features they like.

3. Create a protocol like LSP (language server protocol) but for shells, where the terminal is responsible for getting input, and then uses the protocol to communicate with the shell to get auto-completions (this can be another protocol between the shell and CLI programs), syntax highlighting, and also to have it run the command.
This is similar to how the editor Neovim is split to backends & frontends, where the backend implements the logic and the frontend is the viewport and what the user interacts with.
This can also let the shell render things which are not possible using VT100 codes, but the codes still need to be supported for existing software (that isn't the shell itself) which outputs control codes.

I think an LSP revolution can happen again, I do think OSS shells will co-operate, but it's highly ambitious. A man can dream.

**Reply by Marin-Kitagawa** [reply]
Any update on this? It has been so long

**Comment by zadjii-msft** [comment]
@SuzunaMinami I'd say it's highly unlikely that we explicitly copy the "blocks" style of output as Warp has it. However, there's a lot of work being tracked in #13445 that should accomplish the same goals:

1. Okay, not these. I have a separate plan in mind for these, but Wapr doesn't have these yet either.
2. Jumping bewteen blocks of output - that's possible today with scrollbar marks and shell integration. 
3. Kinda tracked in #1595 
4. This we probably won't be in the business of. I feel like we're better off leaving the management of the input line to the shell, who can actually provide rich context-relevant suggestions, across ssh/vms/whatever, without heuristics. 
5. see 1.
